File: Practica 4/CivitasJava4/CivitasJava/src/civitasjava/Especulador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitasjava;

/**
 *
 * @author juanc
 */
public class Especulador extends Jugador{
    private int fianza;
    private static int factorEspeculador = 2;
    protected final static int CasasMax= 8;
    protected final static int HotelesMax = 4;
    
    public Especulador(Jugador otro, int fianza) {
        super(otro);
        this.fianza = fianza;
        
        if(this.propiedades != null)
            for (int i = 0; i < this.propiedades.size(); i++) 
                this.propiedades.get(i).actualizaPropietarioPorConversion(otro);
        
    }
    
    @Override
    boolean encarcelar(int numCasillaCarcel) {
        this.moverACasilla(numCasillaCarcel); 
     
        if( debeSerEncarcelado() ){
            if ( (this.saldo-this.fianza) < 0 ) {
                this.encarcelado = true;
                Diario.getInstance().ocurreEvento("El jugador " + this.getNombre() 
                                                  + " ha sido encarcelado");
            }
            else {
                this.saldo -= this.fianza;
            }
        }
        
        return this.encarcelado;
    }
    
    @Override
    boolean pagaImpuesto(float cantidad){
        return super.paga(cantidad/2);
    }
    
    @Override
    public String toString(){
     
        String numPropiedades;
        if( this.getPropiedades() != null )
            numPropiedades =  "" + this.propiedades.size();
        else 
            numPropiedades = "0";
        
        
        String info = ( "Nombre: "          + this.getNombre() + " (Especulador)"
              + "\nSaldo: "                 + this.getSaldo()
              + "\nNumero de propiedades "  + numPropiedades
              + "\nCasilla actual: "        + this.getNumCasillaActual()
              + "\nFianza: "                + this.fianza);
        
        return info;
    }
}

File: Practica 4/CivitasRuby4/CivitasRuby/lib/especulador.rb
#encoding:utf-8
require_relative "jugador"

module Civitas
  class Especulador < Jugador
    @@factor_especulador=2

    def initialize(jugador,fianza)
      @fianza = fianza
        
      nuevo_especulador(jugador)
    end
    
    def nuevo_especulador(jugador)
      copia(jugador)
      
      if(@propiedades != nil)
        for i in 0...@propiedades.size
          @propiedades.at(i).actualiza_propietario_por_conversion(jugador)
        end
      end
      
    end
    
    def encarcelar(num_casilla_carcel)
      mover_a_casilla(num_casilla_carcel)
      
      if debe_ser_encarcelado
        if ( @saldo < @fianza ) 
          Diario.instance.ocurre_evento("Se ha encarcelado al especulador #{@nombre}")
          @encarcelado = true
        else
          @saldo = @saldo-@fianza
          Diario.instance.ocurre_evento("No se encarcelo al especulador #{@nombre} porque ha pagado la fianza")
        end
      end
      
      @encarcelado
    end
    
    def paga_impuesto(cantidad)
      super(cantidad/2)
    end
    #Num de props: #{@propiedades.length },\n
    def to_s
      cad = super
      cad = cad + "Fianza: #{@fianza} "

      cad
    end 
    
    def casas_max
      super*@@factor_especulador
    end
    
    def hoteles_max
      super*@@factor_especulador
    end
  end
end
